src/server/bitops_family.cc

```diff
@@ -577,14 +592,513 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
   HandleOpValueResult(res, cntx);
 }
 
+// GCC yields a wrong warning about uninitialized optional use
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+
+enum class EncodingType { UINT, INT, NILL };
+
+struct CommonAttributes {
+  EncodingType type;
+  size_t encoding_bit_size;
+  size_t offset;
+};
+
+// We either return the result of the subcommand (int64_t) or Nill (empty optional)
+// to represent overflow/underflow failures
+using ResultType = std::optional<int64_t>;
+
+struct Overflow {
+  enum Policy { WRAP, SAT, FAIL };
+
+  // Used to check for unsigned overflow/underflow.
+  // If incr is non zero, we check for overflows in the expression incr + *value
+  // If incr is zero, we check for overflows in the expression *value
+  // If the overflow fails because of Policy::FAIL, it returns false. Otherwise, true.
+  // The result of handling the overflow is stored in the pointer value
+  bool UIntOverflow(int64_t incr, size_t total_bits, int64_t* value) const;
+
+  // Used to check for signed overflow/underflow.
+  // If incr is non zero, we check for overflows in the expression incr + *value
+  // If incr is zero, we check for overflows in the expression *value
+  // If the overflow fails because of Policy::FAIL, it returns false. Otherwise, true.
+  // The result of handling the overflow is stored in the pointer value
+  bool IntOverflow(size_t total_bits, int64_t incr, bool add, int64_t* value) const;
+
+  Policy type = WRAP;
+};
+
+bool Overflow::UIntOverflow(int64_t incr, size_t total_bits, int64_t* value) const {
+  // total up to 63 bits -- we do not support 64 bit unsigned
+  const uint64_t max = (1UL << total_bits) - 1;
+
+  uint64_t incr_value = incr;
+  if (incr_value + *value > max) {
+    switch (type) {
+      case Overflow::WRAP:
+        // safe to do, won't overflow, both incr and value are <= than 2^63 - 1
+        *value = (incr_value + *value) % max;
+        break;
+      case Overflow::SAT:
+        *value = max;
+        break;
+      case Overflow::FAIL:
+        *value = 0;
+        return false;
+    }
+    return true;
+  }
+
+  *value = incr_value + *value;
+  return true;
+}
+
+bool Overflow::IntOverflow(size_t total_bits, int64_t incr, bool add, int64_t* value) const {
+  // first check the parsed value is in range
+  const int64_t max = (1UL << total_bits) - 1;
+  const int64_t min = -max - 1;
+  const uint64_t total_elements =
+      total_bits == 63 ? std::numeric_limits<uint64_t>::max() : static_cast<uint64_t>(max) * 2 + 2;
+
+  auto switch_overflow = [&](int64_t wrap_case, int64_t sat_case, int64_t i) {
+    switch (type) {
+      case Overflow::WRAP: {
+        int64_t res = add ? incr - i : *value % total_elements;
+        *value = wrap_case + res;
+        break;
+      }
+      case Overflow::SAT:
+        *value = sat_case;
+        break;
+      case Overflow::FAIL:
+        *value = 0;
+        return false;
+    }
+    return true;
+  };
```

Comment:
Looks really fragile but (almost?) correct to me 🙂 

For example, are you aware that you implicitly cast `*value` to uint64_t? What happens in case of an underflow when *value is negative?

A few DCHECKS wouldn't hurt